### ISSUE TITLE
COPY fails with multiple files when destination folder doesn't exist

### ISSUE THREAD

**Issue by EricHripko** [containers/buildah#1613]
**Description**

**Steps to reproduce the issue:**
1. Create two empty files: `touch file1 file2`
2. Create the following `Dockerfile`
```
FROM scratch
COPY file1 file2 /bin/
```
3. Build an image: `buildah bud .`

**Describe the results you received:**
Image fails to build. See the log output below:
```
STEP 1: FROM scratch
STEP 2: COPY file1 file2 /bin/
error building at STEP "COPY file1 file2 /bin/": destination "/var/lib/containers/storage/overlay/a0d1243fd0bb31904323e9f4306dbb208ed8707a6e14b76db528ccc6901f043a/merged/bin" is not a directory
```

**Describe the results you expected:**
A successful build. See `docker build` output for comparison:
```
Step 1/2 : FROM scratch
 ---> 
Step 2/2 : COPY file1 file2 /bin/
 ---> Using cache
 ---> 71d3bbc649d0
Successfully built 71d3bbc649d0
```

**Output of `rpm -q buildah` or `apt list buildah`:**

```
Built off master
```

**Output of `buildah version`:**

```
Version:         1.9.0-dev
Go Version:      go1.12.4
Image Spec:      1.0.0
Runtime Spec:    1.0.0
CNI Spec:        0.4.0
libcni Version:  
Git Commit:      
Built:           Thu Jan  1 01:00:00 1970
OS/Arch:         linux/amd64
```

**Output of `podman version` if reporting a `podman build` issue:**

```
N/A
```

**Output of `cat /etc/*release`:**

```
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=16.04
DISTRIB_CODENAME=xenial
DISTRIB_DESCRIPTION="Ubuntu 16.04.6 LTS"
NAME="Ubuntu"
VERSION="16.04.6 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.6 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
VERSION_CODENAME=xenial
UBUNTU_CODENAME=xenial
```

**Output of `uname -a`:**

```
Linux scw-35eb3f 4.5.7-docker-4 #1 SMP Tue Jul 12 09:41:14 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
```

**Output of `cat /etc/containers/storage.conf`:**

```
# storage.conf is the configuration file for all tools
# that share the containers/storage libraries
# See man 5 containers-storage.conf for more information

# The "container storage" table contains all of the server options.
[storage]

# Default Storage Driver
driver = "overlay"

# Temporary storage location
runroot = "/var/run/containers/storage"

# Primary read-write location of container storage
graphroot = "/var/lib/containers/storage"

[storage.options]
# AdditionalImageStores is used to pass paths to additional read-only image stores
# Must be comma separated list.
additionalimagestores = [
]

# Size is used to set a maximum size of the container image.  Only supported by
# certain container storage drivers (currently overlay, zfs, vfs, btrfs)
size = ""

# OverrideKernelCheck tells the driver to ignore kernel checks based on kernel version
override_kernel_check = "true"
```

**Comment by EricHripko** [comment]
The problem is that folder doesn't get created by `Builder.Add` earlier, so [this condition](https://github.com/containers/buildah/blob/master/add.go#L142) fails.

I believe the fix here is the same as #1534. I will create a PR this Sunday (no time to finish it today).

**Comment by EricHripko** [comment]
Also, notably - this example doesn't fail on `1.8-dev` version of `buildah` from Apr 13. However, the issue can be triggered with a different example.

**Comment by TomSweeneyRedHat** [comment]
Thanks @EricHripko 
You probably figured this out, but just in case, it seems to work if you use  `FROM alpine` rather than `FROM  scratch`.

**Comment by EricHripko** [comment]
Yep, this is because `/bin` exists in `alpine` image. You can reproduce it on master if you do `COPY ... /my/bin` or to any other path that doesn't exist. `buildah` is supposed to create it - but doesn't (which leads to above-mentioned failures).

**Comment by TomSweeneyRedHat** [comment]
Oh crud, you're right.  I changed the Dockerfile to copy to /baddir instead on alpine and it fails that way too.  Doesn't help for the `FROM scratch` case, but for non-scratch images you can add "RUN mkdir /baddir` to get around it. 

Should be fixed though regardless.

**Comment by EricHripko** [comment]
Investigated this further - it not related to #1534. It's an issue introduced in https://github.com/containers/buildah/commit/c0633e38da4134135d35d32b56194790ca8acecd#diff-65919b07714c5faa770fc0724c4ca886. `securejoin.SecureJoin` drops the trailing slash - which is what [Builder](https://github.com/containers/buildah/blob/master/add.go#L125) uses to determine if target is a folder or a file. I've fixed this by manually restoring the trailing slash if it was present in the original destination spec.

```go
// build.go:495
		if !strings.HasPrefix(finalPath, secureMountPoint) {
			return errors.Wrapf(err, "error resolving copy destination %s", copy.Dest)
		}
		if copy.Dest[len(copy.Dest)-1] == os.PathSeparator {
			finalPath = finalPath + string(os.PathSeparator)
		}
		copy.Dest = strings.TrimPrefix(finalPath, secureMountPoint)
// build.go:501
```

**Comment by nalind** [comment]
I think this is fixed by #1620.

**Comment by EricHripko** [comment]
Yep, works for me on master. Happy to see this fixed, thank you :slightly_smiling_face:
